### DISCUSSION TITLE
[Solved Better] Pi5 - Create3 USB0 Networking

### DISCUSSION THREAD

**Discussion by slowrunner** [iRobotEducation/create3_docs#546] [Ideas]
[The Create3 docs state:](https://iroboteducation.github.io/create3_docs/hw/rpi_hookup/)
> Since the Raspberry Pi® 4 has a USB-C® port capable of an OTG connection, a cable with a USB-C® connector on both ends is all that is required.

But even though the Raspberry Pi 5 also supports the OTG USB0 networking, this certainly is **NOT** all that is required for Pi5 configuration.

The process of getting the USB0 to sport the desired IP for the Create3 connection is [in updated reply below](https://github.com/iRobotEducation/create3_docs/discussions/546#discussioncomment-8531327):

Wali Pi5 and Create3 USB0 Networking Plan:

1)  Ensure WaliP5 is using 5GHz WiFi   
2)  Setup USB0 Network on WaLiPi5              
3)  Setup ntp server on WaLiPi5-PiOS on USB network 
4)  Setup Create3 to use USB Network 
5)  Check Create3 WebServer Logs/About Access, ntp working
6)  Setup Discovery Server to start in WaLiPi5 Dockered ROS
7)  Configure WaliPi5 nodes to be "SuperClients"
8)  Confirm Create3 topics on WaliPi5 (/dock_status, /odom, /ir_intensity)
9)  Confirm WaliPi5 topics seen by Create3 (/cmd_vel, /dock)
10) Confirm Create3 topics reach desktop (/battery_state, /dock_status) and rviz (/odom)
11) Confirm WaliPi5 topics reach rviz (/mapUpdate, /oak/color/image_raw, /scan)
12) Confirm Desktop topics reach Create3 (/dock)

**Comment by slowrunner** [comment]


**Reply by brianabouchard** [reply]
Hello! Funny enough, I had a reply half written to you yesterday and was then pulled into meetings and didn't hit send. I'm glad you tagged me. Admittedly, I have not had time to continue testing the usb0 connection, but my initial attempts were unsuccessful (though not as thorough as I would like - I'm busy this semester teaching with the Pi4 so I haven't had as much time to explore the Pi5 yet), so I'm excited to hear you were able to get it up and operational. Did you do anything special or just followed the Pi 4 instructions (better than I did :))? I'd love to replicate it on my end so I can do soome further testing and see if I can reproduce the behavior you are seeing. 

One thing I noted is that the USB0 networking feature is no longer anywhere in the Raspberry Pi docs, or at least it wasn't back when I was testing this fall. Seems odd and made me wonder if they were stopping supporting it.

**Comment by slowrunner** [comment]
# ISSUES FROM ENABLING USB0 ON PI5 [SOLVED ALL NOW]

I've had to slay a few issues resulting from enabling USB0 networking on the Pi5, and JUST DISCOVERED A BIG, BIG ONE!

BIG ISSUE:  USB0 config on Pi5 prevents RTC wakeup
I noticed last night when I put Wali to "nap" that the Pi5 did not go to drawing 0 Amps / 0 Watts as in prior nights:
![Wali_Nap_With_USB0_Enabled](https://github.com/iRobotEducation/create3_docs/assets/8950202/0719d8e7-915f-49ba-bf93-a349b8669457)

I've just tested putting Wali to sleep for 12 minutes and he didn't wake up.  There is a line in the USB0 enable stuff that hints at this very power number:
```
echo "create3-walipi5 USB Device" > strings/0x409/product
mkdir -p configs/c.1/strings/0x409
echo "CDC" > configs/c.1/strings/0x409/configuration
echo 250 > configs/c.1/MaxPower                       <<<---- ???  250mA  Available Over USB0 ???
echo 0x80 > configs/c.1/bmAttributes
```

This RTC wakeup alarm on the Pi5 is a totally new feature, and for all I can tell the Pi5 USB0 "Host/Gadget Mode" is not officially released, so my Create3 is really venturing in uncharted territory. 

ISSUES SOLVED:
- USB0 Networking changed my USB Audio speaker from device 2,0 to device 0,0 (I don't know how to configure a default setup) 
- USB0 Networking slows network startup so my IP audio announcement needed a delay.  It was announcing "no IP found" to early.

**Reply by slowrunner** [reply]
Figured this one - the ```sudo rpi-update``` reset the POWER_OFF_ON_HALT=1 to 0.  After re-editing the eeprom entry - now working.

**Comment by slowrunner** [comment]
# UPDATE - THIS WORKS WELL

( @brianabouchard  )

# Set up usb0 networking on Raspberry Pi 5 USB-C connector to connect with Create3

REF: https://forums.raspberrypi.com/viewtopic.php?p=2195159#p2195179  
(For a never booted system use REF: https://forums.raspberrypi.com/viewtopic.php?t=364247 for step 1)  

1) sudo nano /etc/network/interfaces.d/g_ether  

```
auto usb0 
allow-hotplug usb0 
iface usb0 inet static
        address 192.168.186.3
        netmask 255.255.0.0

auto usb0.1
allow-hotplug usb0.1
iface usb0.1 inet dhcp

EOF
```

2) sudo apt update && sudo apt upgrade -y

3) add to /boot/firmware/config.txt  
```
dtoverlay2,dr_mode=peripheral
```

4) add to end of /boot/firmware/cmdline.tx  
```
modules-load=dwc2,g_ether
```

5) reboot  

6) ifconfig  
```
docker0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 172.17.0.1  netmask 255.255.0.0  broadcast 172.17.255.255
        ether 02:42:33:ed:b6:4a  txqueuelen 0  (Ethernet)

eth0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        ether d8:3a:dd:b6:36:a1  txqueuelen 1000  (Ethernet)

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        loop  txqueuelen 1000  (Local Loopback)

usb0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.186.3  netmask 255.255.0.0  broadcast 192.168.255.255
        ether ea:78:3a:fa:b6:c5  txqueuelen 1000  (Ethernet)

usb0.1: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        ether ea:78:3a:fa:b6:c5  txqueuelen 1000  (Ethernet)

wlan0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet x.x.x.wifi  netmask 255.255.255.0  broadcast x.x.x.255
        ether d8:3a:dd:b6:36:a2  txqueuelen 1000  (Ethernet)s 0
```

**Reply by brianabouchard** [reply]
Just a quick note for anyone following along on this thread...Instruction 1 is specific to Raspberry Pi OS Bookworm. If you are trying to use Ubuntu 24.04 server with Jazzy and the newly released Create 3 Iron Firmware, Ubuntu 24.04 server uses netplan for networking instead. As such, you can follow step 8 on the Humble install page to configure the network-config executable to replace what is written above under step 1. If you've already booted, you can edit the file `/etc/netplan/50-cloud-init.yaml` to add the usb0 configuration. 

Lastly, for those of you that may have imaged your SD card prior to the last few months, there was an issue with the rpi-eeprom release that was preventing this mode from functioning properly. You can resolve the issue by running `sudo rpi-eeprom-update -a` and rebooting. As of this morning, I was able to use the RPi Imager to write the Ubuntu Server 24.04 image to my SD and did not need to run this command to get it working properly. 

Thanks @slowrunner for all your troubleshooting and documentation of this! It got me very close to where I needed to be and would have been much harder otherwise.

**Comment by slowrunner** [comment]
Working well
